### ISSUE TITLE
memkind_interleave/memkind_hbw_interleave are available but malloc returns nil 

### ISSUE THREAD

**Issue by bcernohous** [memkind/memkind#18]
On SLES12 

[hbw_interleave.zip](https://github.com/memkind/memkind/files/266876/hbw_interleave.zip)

$ ./hbw_interleave

 memkind_hbw_interleave          : is available(yes), total 016942243840, free 016874237952, mask 0X0002 , malloc'd ptrs (nil)/(nil)
                                 : unable to memkind_malloc() from available kind

 memkind_interleave              : is available(yes), total 118363762688, free 115329839104, mask 0X0003 , malloc'd ptrs (nil)/(nil)
                                 : unable to memkind_malloc() from available kind

$ numactl --hardware 
available: 2 nodes (0-1)
..
node 0 size: 96723 MB
node 0 free: 93933 MB
node 1 size: 16157 MB
node 1 free: 16092 MB
node distances:
node   0   1 
  0:  10  31 
  1:  31  10

**Comment by Kkulakowski** [comment]
Hi Bob,

I tried your code and I think I found the issue. You are using:

`
            kind->ops->get_mbind_nodemask(kind, &nodemask,
                                          NUMA_NUM_NODES);
`

Where nodemask is:

`    unsigned long nodemask = 0;`

and `NUMA_NUM_NODES` on x86_64 is equal 128. This basically means that numa functions called by memkind will assume that size of `nodemask` is 128 (where in fact it is sizeof(unsigned long) == 64). This result in stack corruption which overrides one of pointers value.

Here is your code with my fix:
[hbw_interleave_fixed.zip](https://github.com/memkind/memkind/files/277747/hbw_interleave_fixed.zip)

I changed `    unsigned long nodemask = 0;` to `    nodemask_t nodemask;`. It's also worth noticing that I'm using latest (today published) version of memkind, so you might need to remove `#include <memkind/internal/memkind_private.h>` before compilation.

Please let me know if that is solving your issue.
Regards,
Krzysztof.

**Comment by bcernohous** [comment]
That fixed code link is 404 for me.  Can you send me the zip as an attachment or .c?

I did, at some point, notice the 64/128 problem.  Thought I fixed it but I’ll have to look around.  I’ve been very confused about all the const’s and structs.  Bitmasks vs nodemasks vs long.

You’re api said unsigned long so I think my fix was to use numa_max_possible_node() for maxnode, which is 63(?) rather than stick with 128 and switch to nodemask.

int (* get_mbind_nodemask)(struct memkind *kind, unsigned long *nodemask, unsigned long maxnode);

I saw the v1.1.0 email but I haven’t downloaded anything.  Do we lose access to the ops table?  We’d like to be able to get the nodemasks that you use, like that code demonstrates.  Have you made all that private now?

**Comment by Kkulakowski** [comment]
Hi, I checked and I'm able to download this code on my mobile so I believe it might be some problem with you browser or something similar. Here is direct link:

https://github.com/memkind/memkind/files/277747/hbw_interleave_fixed.zip

I know it might be tricky, however this was the best solution we had to avoid using non-standard types. I'm hoping that after you take a look at my code it would all be much simpler :)

Speaking of ops. For now they have been moved to this memkind/internal/memkind_private.h, so they are still accessible. We want to solve few design issues we identified and memkind struct definition is part of that problem. 

I hear you feedback, and I think that getting kind nodemask will be possible, although It may require some code changes in future (that was the message we wanted to send by moving this struct to _private header).

**Comment by bcernohous** [comment]
That’s interesting.  The tab was still open on the 404 and I refreshed it and it worked.   Maybe a github hiccup.

I looked at my newer code (the test I sent you was older memkind testing I had done) and I use a combination.  Basically I use a nodemask and cast it to unsigned long on your api AND use numa_max_possible_node() since that seemed a reasonable value for maxnode :

kind->ops->get_mbind_nodemask(kind, (unsigned long *)&nodemask, numa_max_possible_node());

I think that works ok.  I’ll have to verify it a little more carefully but I don’t have any dev system available right now.

I’ll look at your fix too.

For now, we’ll use the private interface and keep an eye on your updates.  We want the nodemask because we promised a customer we’d (SHMEM) cooperate and standardize with Intel on how to use MCDRAM.  Part of that is agreeing on the node config so we want use memkind for config and we use nodemask for other things too.
